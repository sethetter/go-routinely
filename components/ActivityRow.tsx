import * as React from 'react'
import * as moment from 'moment'

import ActivityDay from './ActivityDay'

export interface ActivityRowProps {
  activity: Activity
  weekDays: Date[]
  logsForActivity: ActivityLog[]
  createLog: (params: Partial<ActivityLog>) => void
}

const ActivityRow = ({
  activity,
  weekDays,
  logsForActivity,
  createLog,
}: ActivityRowProps) => {
  const activityDays = weekDays.map(d => {
    const logsForDay = logsForActivity.filter(l =>
      moment(l.occurredAt).isSame(d, 'day')
    )

    return (
      <ActivityDay
        key={d.valueOf()}
        day={d}
        activityId={activity._id}
        logsForDay={logsForDay}
      />
    )
  })

  const createLogForNow = () => {
    createLog({ activityId: activity._id })
  }

  return (
    <tr key={activity._id} className="ActivityRow">
      <td onClick={createLogForNow} className="activity-name">{activity.name}</td>
      {activityDays}
    </tr>
  )
}

export default ActivityRow
