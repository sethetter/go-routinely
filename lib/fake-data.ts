import * as moment from 'moment'

interface AppData {
  startOfWeek: Date
  activities: Activity[]
  activityLogs: ActivityLog[]
  weekDays: Date[]
}

const startOfWeek = new Date('03/25/2018')

const WEEKDAYS = [
  moment(startOfWeek).day('Sunday').startOf('day').toDate(),
  moment(startOfWeek).day('Monday').startOf('day').toDate(),
  moment(startOfWeek).day('Tuesday').startOf('day').toDate(),
  moment(startOfWeek).day('Wednesday').startOf('day').toDate(),
  moment(startOfWeek).day('Thursday').startOf('day').toDate(),
  moment(startOfWeek).day('Friday').startOf('day').toDate(),
  moment(startOfWeek).day('Saturday').startOf('day').toDate(),
]

const DATA: AppData = {
  startOfWeek,
  activities: [
    { id: 1, name: 'Brush teeth', value: 1 },
    { id: 2, name: 'Taekwondo', value: 2 },
    { id: 3, name: 'Clean room', value: 3 },
  ],
  activityLogs: [
    { id: 1, name: 'Brush teeth', value: 1, activityId: 1, occurredAt: WEEKDAYS[0] },
    { id: 2, name: 'Brush teeth', value: 1, activityId: 1, occurredAt: WEEKDAYS[0] },
    { id: 3, name: 'Taekwondo', value: 2, activityId: 2, occurredAt: WEEKDAYS[3] },
    { id: 4, name: 'Clean room', value: 3, activityId: 3, occurredAt: WEEKDAYS[2] },
  ],
  weekDays: WEEKDAYS
}

export const getFakeData = (): AppData => Object.assign({}, DATA)