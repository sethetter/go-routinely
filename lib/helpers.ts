import * as moment from 'moment'

export function startOfWeekFromDate (date: Date): Date {
  return moment(date).day('Sunday').startOf('day').toDate()
}

export function weekDaysFromDate (date: Date): Date[] {
  // Subtract one to start so loop can always add one
  const startOfWeek = moment(date).day('Sunday').subtract(1, 'day').startOf('day')
  const weekDays = []

  for (let i = 0; i < 7; i++) {
    weekDays.push(startOfWeek.add(1, 'day').toDate())
  }

  return weekDays
}
