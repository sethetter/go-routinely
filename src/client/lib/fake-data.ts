import moment from 'moment'

interface AppData {
  user: UserData
  startOfWeek: Date
  activities: Activity[]
  activityLogs: ActivityLog[]
  weekDays: Date[]
}

const startOfWeek = new Date('03/25/2018')

const WEEKDAYS = [
  moment(startOfWeek).day('Sunday').startOf('day').toDate(),
  moment(startOfWeek).day('Monday').startOf('day').toDate(),
  moment(startOfWeek).day('Tuesday').startOf('day').toDate(),
  moment(startOfWeek).day('Wednesday').startOf('day').toDate(),
  moment(startOfWeek).day('Thursday').startOf('day').toDate(),
  moment(startOfWeek).day('Friday').startOf('day').toDate(),
  moment(startOfWeek).day('Saturday').startOf('day').toDate(),
]

const DATA: AppData = {
  startOfWeek,
  user: { id: 'aosidfjoasidfj' },
  activities: [
    { _id: '1', name: 'Brush teeth', value: 1 },
    { _id: '2', name: 'Taekwondo', value: 2 },
    { _id: '3', name: 'Clean room', value: 3 },
  ],
  activityLogs: [
    { _id: '1', name: 'Brush teeth', value: 1, activityId: '1', occurredAt: WEEKDAYS[0] },
    { _id: '2', name: 'Brush teeth', value: 1, activityId: '1', occurredAt: WEEKDAYS[0] },
    { _id: '3', name: 'Taekwondo', value: 2, activityId: '2', occurredAt: WEEKDAYS[3] },
    { _id: '4', name: 'Clean room', value: 3, activityId: '3', occurredAt: WEEKDAYS[2] },
  ],
  weekDays: WEEKDAYS
}

export const getFakeData = (): AppData => Object.assign({}, DATA)
