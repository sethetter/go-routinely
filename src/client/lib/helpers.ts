import * as moment from 'moment'

export function startOfWeekFromDate (date: Date): Date {
  return moment(middleOfDay(date)).day('Sunday').toDate()
}

export function weekDaysFromDate (date: Date): Date[] {
  // Subtract one to start so loop can always add one
  const startOfWeek = moment(middleOfDay(date)).subtract(1, 'day')
  const weekDays = []

  for (let i = 0; i < 7; i++) {
    weekDays.push(startOfWeek.add(1, 'day').toDate())
  }

  return weekDays
}

export function middleOfDay (date: Date): Date {
  return moment.utc(date).startOf('day').add(12, 'hours').toDate()
}
