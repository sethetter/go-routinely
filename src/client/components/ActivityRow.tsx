import * as React from 'react'
import moment from 'moment'

import { middleOfDay } from '../lib/helpers'

import ActivityDay from './ActivityDay'

export interface ActivityRowProps {
  activity: Activity
  weekDays: Date[]
  logsForActivity: ActivityLog[]
  createLog: (params: Partial<ActivityLog>) => void
}

const ActivityRow = ({
  activity,
  weekDays,
  logsForActivity,
  createLog,
}: ActivityRowProps) => {
  const activityDays = weekDays.map(d => {
    const logsForDay = logsForActivity.filter(l =>
      moment.utc(l.completedAt).isSame(d, 'day')
    )

    return (
      <ActivityDay
        key={d.valueOf()}
        day={d}
        activityId={activity._id}
        logsForDay={logsForDay}
        createLog={createLog}
      />
    )
  })

  const createLogForNow = () => {
    createLog({
      activityId: activity._id,
      completedAt: middleOfDay(new Date()),
    })
  }

  return (
    <tr key={activity._id} className="ActivityRow">
      <td onClick={createLogForNow} className="activity-name">{activity.name}</td>
      {activityDays}
    </tr>
  )
}

export default ActivityRow
