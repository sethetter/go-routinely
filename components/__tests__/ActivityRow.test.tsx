import * as React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import { weekDaysFromDate } from '../helpers'
import ActivityRow, { ActivityRowProps } from './ActivityRow'

import { getFakeData } from '../fake-data'

const DATA = getFakeData()

const activity = DATA.activities[0]
const props: ActivityRowProps = {
  activity,
  logsForActivity: DATA.activityLogs.filter(l => {
    return l.activityId === activity.id
  }),
  weekDays: weekDaysFromDate(new Date())
}

describe('<ActivityRow />', () => {
  it('renders successfully', () => {
    const component = shallow(<ActivityRow {...props} />)

    expect(shallowToJson(component)).toMatchSnapshot()
  })
})
