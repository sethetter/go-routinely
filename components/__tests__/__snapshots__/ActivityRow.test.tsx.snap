// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActivityRow /> renders successfully 1`] = `
<tr
  className="ActivityRow"
  key="1"
>
  <td>
    Brush teeth
  </td>
  <ActivityDay
    activityId={1}
    day={2018-05-06T05:00:00.000Z}
    key="1525582800000"
    logsForDay={
      Array [
        Object {
          "activityId": 1,
          "id": 1,
          "name": "Brush teeth",
          "occurredAt": 2018-05-06T05:00:00.000Z,
          "value": 1,
        },
        Object {
          "activityId": 1,
          "id": 2,
          "name": "Brush teeth",
          "occurredAt": 2018-05-06T05:00:00.000Z,
          "value": 1,
        },
      ]
    }
  />
  <ActivityDay
    activityId={1}
    day={2018-05-07T05:00:00.000Z}
    key="1525669200000"
    logsForDay={Array []}
  />
  <ActivityDay
    activityId={1}
    day={2018-05-08T05:00:00.000Z}
    key="1525755600000"
    logsForDay={Array []}
  />
  <ActivityDay
    activityId={1}
    day={2018-05-09T05:00:00.000Z}
    key="1525842000000"
    logsForDay={Array []}
  />
  <ActivityDay
    activityId={1}
    day={2018-05-10T05:00:00.000Z}
    key="1525928400000"
    logsForDay={Array []}
  />
  <ActivityDay
    activityId={1}
    day={2018-05-11T05:00:00.000Z}
    key="1526014800000"
    logsForDay={Array []}
  />
  <ActivityDay
    activityId={1}
    day={2018-05-12T05:00:00.000Z}
    key="1526101200000"
    logsForDay={Array []}
  />
</tr>
`;
